libr/bin/format/mdmp/mdmp.c

```diff
@@ -1,371 +1,265 @@
+		for (i = 0; i < handle_operation_list->number_of_entries; i++) {
+			handle_operations = (struct avrf_handle_operation *)((ut8 *)handle_operation_list + sizeof (struct minidump_handle_operation_list));
+			r_list_append (obj->streams.operations, &(handle_operations[i]));
+		}
+
 		break;
-	case TokenStream:
+	case LAST_RESERVED_STREAM:
+		eprintf ("TODO: Parse last reserved stream!\n");
 		break;
-	case JavaScriptDataStream:
+	case UNUSED_STREAM:
+	case RESERVED_STREAM_0:
+	case RESERVED_STREAM_1:
+		/* Silently ignore reserved streams */
 		break;
-	default
-			:
-		eprintf("Warning in r_bin_mdmp_init_directory: uknown stream %d\n", dir->StreamType);
+	default:
+		eprintf ("WARNING: Invalid or unsupported enumeration encountered %i\n", entry->stream_type);
+		return false;
+	}
+	return true;
+}
+
+static bool r_bin_mdmp_init_directory(struct r_bin_mdmp_obj *obj) {
+	int i;
+	ut8 *directory_base;
+	struct minidump_directory *entry;
+
+	directory_base = obj->b->buf + obj->hdr->stream_directory_rva;
+
+	/* Parse each entry in the directory */
+	for (i = 0; i < obj->hdr->number_of_streams; i++) {
+		entry = (struct minidump_directory *)(directory_base + (i * sizeof (struct minidump_directory)));
+		r_bin_mdmp_init_directory_entry (obj, entry);
+	}
+
+	return true;
+}
+
+static int r_bin_mdmp_init(struct r_bin_mdmp_obj *obj) {
+	if (!r_bin_mdmp_init_hdr (obj)) {
+		eprintf ("Error: Failed to initialise header\n");
+		return false;
+	}
+
+	if (!r_bin_mdmp_init_directory (obj)) {
+		eprintf ("Error: Failed to initialise directory structures!\n");
+		return false;
 	}
+
 	return true;
 }
 
-PMINIDUMP_STRING r_bin_mdmp_locate_string(struct r_bin_mdmp_obj *obj, RVA Rva) {
-	if (Rva >= obj->b->length)
+void r_bin_mdmp_free(struct r_bin_mdmp_obj *obj) {
+	if (!obj) {
+		return;
+	}
+
+	if (obj->streams.ex_threads) r_list_free (obj->streams.ex_threads);
+	if (obj->streams.memories) r_list_free (obj->streams.memories);
+	if (obj->streams.memories64.memories) r_list_free (obj->streams.memories64.memories);
+	if (obj->streams.memory_infos) r_list_free (obj->streams.memory_infos);
+	if (obj->streams.modules) r_list_free (obj->streams.modules);
+	if (obj->streams.operations) r_list_free (obj->streams.operations);
+	if (obj->streams.thread_infos) r_list_free (obj->streams.thread_infos);
+	if (obj->streams.threads) r_list_free (obj->streams.threads);
+	if (obj->streams.unloaded_modules) r_list_free (obj->streams.unloaded_modules);
+
+	if (obj->kv) {
+		sdb_free (obj->kv);
+		obj->kv = NULL;
+	}
+	if (obj->b) {
+		r_buf_free (obj->b);
+		obj->kv = NULL;
+	}
+	R_FREE (obj);
+
+	return;
+}
+
+struct r_bin_mdmp_obj *r_bin_mdmp_new_buf(struct r_buf_t *buf) {
+	bool streams;
+	struct r_bin_mdmp_obj *obj;
+
+	obj = R_NEW0 (struct r_bin_mdmp_obj);
+	obj->kv = sdb_new0 ();
+	obj->b = r_buf_new ();
+	obj->size = (ut32)buf->length;
+
+	streams = true;
+	if (!(obj->streams.ex_threads = r_list_new ())) streams = false;
```

Comment:
this seems like a goto emulation what about doing like this?

what about sdoing it like:
```
bool fail = false;
fail |= !(obj->streamds.memories = r_list_new());
if (fail) {...}
```